beiboot: feature parity with cockpit-ssh

martinpitt opened the pull request:
In order to use cockpit-beiboot as cockpit-ssh replacement from the "normal" (not Client mode) login page, it needs to consider the given username and password. It also needs to properly handle "unknown host key" prompts.

This paves the way for completely replacing cockpit-ssh with cockpit.beiboot. See https://issues.redhat.com/browse/COCKPIT-954 and #19441.

 - [x] Builds on top of PR #19413
 - [x] Preferably fix PR #19414
 - [x] Builds on top of PR #19859
 - [x] Builds on top of PR #20967

martinpitt commented:
@allisonkarlitskaya This is still draft. I wanted to get something working, so that it becomes easier (and more enjoyable) to iterate on it. I know you had something else in mind, like changing the ferny API instead of doing things like `'password:' in prompt` on the cockpit side. Nevertheless, I'd appreciate a quick first opinion about the direction. Thanks!

allisonkarlitskaya commented on src/cockpit/beiboot.py:
```diff
@@ -186,9 +194,19 @@ def __init__(self, router: Router, destination: str, args: argparse.Namespace):
+        if auth is not None and auth.startswith('Basic '):
+            user, basic_password = base64.b64decode(auth.split(' ', 1)[1]).decode().split(':', 1)
+            logger.debug("got username %s and password from Basic auth", user)
+            ssh_opts = ['-l', user]
```
I guess this overrides `user@` given as part of the host specification on the command line?

martinpitt replied:
Yes,  it does. AFAICS there can only ever be one or the other -- "client" mode doesn't have "Username:" and "Password:" fields (and thus no basic auth), and "standard" login page never supported adding a user name in the "Connect to:" field. My gut feeling is that "Username:" should win, but if you prefer I can add a check for `@` in `self.destination` and either handle it or error out with "make up your mind"

martinpitt replied:
There is a test for this now, so resolving.

allisonkarlitskaya commented on test/verify/check-static-login:
```diff
@@ -955,6 +955,68 @@ matchrule = <SUBJECT>^DC=LAN,DC=COCKPIT,CN=alice$
+    @testlib.skipImage("needs pybridge", "rhel-8*", "centos-8*")
+    # enable this once our cockpit/ws container can beiboot
+    @testlib.skipOstree("client setup does not work with ws container")
+    def testLoginSshBeiboot(self):
```
Can you add a test scenario to find out what happens if you specify a username *and* a same/different `user@host`?

martinpitt replied:
You mean writing `user@host` into the `Connect to:` field? As I said above, that has never been intended to work, but I'll add it to pin down the behaviour.

martinpitt replied:
Done.

allisonkarlitskaya commented on src/cockpit/beiboot.py:
```diff
@@ -316,8 +315,11 @@ async def run(args) -> None:
     except ferny.InteractionError as exc:
-        sys.exit(str(exc))
+        logger.debug("ferny.InteractionError: %s", exc)
+        bridge.write_control(command='init', problem='authentication-failed', message=str(exc))
```
Nice.  Thanks.

allisonkarlitskaya reviewed (changes requested):
This looks like it's coming together really nicely and I'm starting to get excited about the finished result.  :rocket:

martinpitt commented on selinux/HACKING.md:
```diff
@@ -0,0 +1,11 @@
+# Changing cockpit's SELinux policy
+
+The clean way to do that is to edit the policy files, and rebuilding the rpms
+and image with e.g. `test/image-prepare -q fedora-XX`.
```
Thanks! Applied.

martinpitt commented on src/cockpit/beiboot.py:
```diff
@@ -50,7 +53,14 @@ def ensure_ferny_askpass() -> Path:
     if xdg_cache_home is None:
         xdg_cache_home = os.path.expanduser('~/.cache')
-    os.makedirs(xdg_cache_home, exist_ok=True)
+    try:
```
Ah, well spotted -- this was a quick "get out of my way!" hack, but indeed it should rather use the installed binary. Replaced with a new commit that does that.

martinpitt commented on src/cockpit/beiboot.py:
```diff
@@ -134,10 +134,18 @@ class AuthorizeResponder(ferny.InteractionResponder):
+        self.basic_password = basic_password
 
     async def do_askpass(self, messages: str, prompt: str, hint: str) -> Optional[str]:
+        if self.basic_password and 'password:' in prompt.lower():
```
Yep, see my [comment above](https://github.com/cockpit-project/cockpit/pull/19401#issuecomment-1736846236), that's why it's still draft. Of course we could also consider/mark this as `# HACK` and update ferny as a follow-up. I still *do* want to do that, but it's much nicer to iterate on something that actually works.

martinpitt commented on src/cockpit/beiboot.py:
```diff
@@ -186,9 +194,19 @@ def __init__(self, router: Router, destination: str, args: argparse.Namespace):
 
     async def do_connect_transport(self) -> None:
+        # do we have user/password (Basic auth) from the login page?
+        auth = await self.router.request_authorization("*")
```
Not a strong one. The biggest one is to limit the scope of the `user`, `password` etc. variables, to keep the whole initialization together. But I'm happy to move this out somewhere if you prefer.

martinpitt commented on test/verify/check-static-login:
```diff
@@ -955,6 +955,68 @@ matchrule = <SUBJECT>^DC=LAN,DC=COCKPIT,CN=alice$
+
+        m.write("/etc/cockpit/cockpit.conf", f"""
+[Ssh-Login]
+Command = {self.libexecdir}/cockpit-beiboot
```
I hope that your PR wins :grin: I think this one will probably require more rounds.

martinpitt replied:
It landed. Done locally, will push tomorrow once I figured out the remaining corner cases.

martinpitt commented on selinux/cockpit.te:
```diff
@@ -49,8 +49,15 @@ can_exec(cockpit_ws_t,cockpit_ws_exec_t)
+    type ssh_exec_t;
+')
+can_exec(cockpit_ws_t,ssh_exec_t)
+# cockpit-client-askpass is written as temporary file
+can_exec(cockpit_ws_t,cockpit_tmp_t)
```
I think we can take this out again. We now expect the packaged cockpit-beiboot to call the packaged cockpit-askpass, and for the "remote beiboot" case, everything is just in temp files, and cockpit-ws isn't involved.

martinpitt replied:
Done.

martinpitt commented:
Uh, [interesting SELinux rejection](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19401-20230927-103256-ee489a13-fedora-39-other/log.html#172):
```
audit: type=1400 audit(1695811735.041:4956): avc:  denied  { read } for  pid=18477 comm="cockpit-beiboot" name="possible" dev="sysfs" ino=42 scontext=system_u:system_r:cockpit_ws_t:s0 tcontext=system_u:object_r:sysfs_t:s0 tclass=file permissive=0
```

I have no idea what the "possible" file is, we don't access that anywhere in cockpit, systemd_ctypes, beipack, or ferny. There's this:
```
./devices/system/cpu/possible
./devices/system/node/possible
```

martinpitt commented:
@allisonkarlitskaya I'd like to ask you for another round, and your opinion:

 - Go ahead with this, and use it to model what we want to change in ferny API (in particular, move all the string parsing there)
 - Drop the topmost commit (Improve beiboot error reporting for fatal login failures), rework ferny for throwing more fine-grained `InteractionError` subclasses, such as `AuthenticationFailure` or `UnknownHostError` 
 - Completely block this on ferny API rework

martinpitt commented:
Beiboot is broken [in an interesting way on ubuntu-2204](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19401-20230928-054506-d7f3fa3e-ubuntu-2204-other/log.html#161). I poked at this for half an hour, but I think I'd like to confer with @allisonkarlitskaya about that -- also, let's handle that in a separate PR.

This also got [more of these "execution context destroyed" flakes](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19401-20230928-054506-d7f3fa3e-fedora-38-firefox-other/log.html#172). It is really hard to reproduce locally, I catch it in maybe 1 out of 20 runs. I have an idea how to fix it, and so far it passed locally. Let the bots tell me wrong :grin:

martinpitt commented:
As this is getting too big, I split out the harmless bits into PR #19413, and the hairy EPIPE crash into PR #19414 -- let's talk about them separately.

martinpitt commented:
I restructured this further to also add the host key message processing. That captures all of the SSH message processing that we'll have to do, and takes beiboot logic changes out of #19441.

martinpitt commented:
The "failed" TF test is wrong, see https://github.com/packit/packit-service/issues/2216 . Ignore.

allisonkarlitskaya commented on src/cockpit/beiboot.py:
```diff
@@ -332,8 +331,18 @@ async def run(args) -> None:
+        msg = str(exc)
+        if 'Permission denied' in msg:
+            problem = 'authentication-failed'
+        elif 'resolve host' in msg:
```
There's already code inside of ferny to do this and we should definitely use it.

martinpitt replied:
The `Session` API [does this](https://github.com/allisonkarlitskaya/ferny/blob/main/src/ferny/session.py#L160), [get_exception_for_ssh_stderr](https://github.com/allisonkarlitskaya/ferny/blob/main/src/ferny/errors.py#L102)

martinpitt replied:
Converted to the ferny.errors API.

allisonkarlitskaya commented on src/cockpit/beiboot.py:
```diff
@@ -165,12 +166,21 @@ async def do_askpass(self, messages: str, prompt: str, hint: str) -> Optional[st
+        fp_match = re.search(r'\n(\w+) key fingerprint is ([^.]+)\.', prompt)
+        args = {}
+        if fp_match:
+            args['host-key'] = f'{fp_match.group(2)} {fp_match.group(1)}'
```
I guess this isn't quite right — this is the fingerprint, not the key...

martinpitt replied:
Hm yes, but we don't ever get the actual key via SSH_ASKPASS; I figure that's why you went through this whole business of `KnownHostCommand`? Do we need to use that after all?

martinpitt replied:
I.e. I just consider that as the "new key format" -- key fingerprints are just the "language" that `ssh` speaks, while for `cockpit-ssh` it was the full keys. But does that matter? Or what would you suggest?

allisonkarlitskaya replied:
I guess this is still unresolved — I worry about existing users who might be concerned that their key no longer matches, on upgrade.  Do we have a test for that?

allisonkarlitskaya replied:
(don't get me wrong — I think this change is good and ultimately necessary if we're to abandon our dependency on hostkey commands... but I feel like we need to try a bit harder on the migration front...)

martinpitt replied:
The ultimate/persistent place to store that information is ~/.ssh/known_hosts, both for cockpit-ssh and cockpit-beiboot. That format is fixed, and ssh/cockpit-ssh always store the full keys there. As far as I can see this only affects the wire protocol between cockpit-{ssh,beiboot} and the UI.

What migration would/could we possibly do here? To be honest I don't even have an idea how to "test" this against the old format -- we either use cockpit-ssh or cockpit-beiboot, and the UI just needs to get along with either. We never showed full keys in the UI, always just fingerprints.

martinpitt replied:
It's been a while, so need to page that stuff back into my brain. Login page currently stores the full host key in localStorage:
```
known_hosts:"{"[127.0.0.2]:2201":"[127.0.0.2]:2201 ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBBERg5Z5mVTMkmOaIICuFnrLZs/DWPFq/PvNkrady3sxvETrtyF58yjYfIt9iOL6Tcve65F+iDYUsyv9r431ToM=\n"}"
```

The corresponding fingerprint is "SHA256:sq4xrPrvIQ7MChHi8fZedhXQxqqba3BDLsuSOu0/SPc"

It's awkward to call `ssh-keygen` on this on the login page, as at that point we don't even have a bridge. But that fingerprint calculation isn't obvious -- the most straightforward computation gives something else:

```
# echo -n 'AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBBERg5Z5mVTMkmOaIICuFnrLZs/DWPFq/PvNkrady3sxvETrtyF58yjYfIt9iOL6Tcve65F+iDYUsyv9r431ToM=' | base64 -d | sha256sum -
b2ae31acfaef210ecc0a11e2f1f65e7615d0c6aa9b6b70432ecb923aed3f48f7  -
```

I also tried to directly SHA the base64 encoded key, but it's also different.

I found a [stackexchange question](https://security.stackexchange.com/questions/152428/how-is-an-ssh-public-key-fingerprint-calculated), and re-tried with Python and that *almost* works:
```
>>> import hashlib, base64
>>> b64pubkey = 'AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBBERg5Z5mVTMkmOaIICuFnrLZs/DWPFq/PvNkrady3sxvETrtyF58yjYfIt9iOL6Tcve65F+iDYUsyv9r431ToM='
>>> s = hashlib.sha256()
>>> s.update(base64.b64decode(b64pubkey))
>>> h = s.digest()
>>> h
b'\xb2\xae1\xac\xfa\xef!\x0e\xcc\n\x11\xe2\xf1\xf6^v\x15\xd0\xc6\xaa\x9bkpC.\xcb\x92:\xed?H\xf7'
>>> base64.b64encode(h)
b'sq4xrPrvIQ7MChHi8fZedhXQxqqba3BDLsuSOu0/SPc='
```

The difference is the extra `=` padding at the end. We could ignore these for the comparison.

martinpitt replied:
If the key type doesn't match, show the "unknown host", not the "warning danger host key changed" dialog.

martinpitt replied:
I reworked this. The integration test makes relatively strong assumptions, let's see how far I get with that. I may have to weaken it a bit (make more dynamic).

martinpitt replied:
This is mostly working now, everything uses full host keys. See the details and caveat in the commit message.

martinpitt commented:
[curious unit test flake](https://github.com/cockpit-project/cockpit/actions/runs/6430433450/job/17461393055?pr=19401#step:4:485). All tests passed, but at shutdown:
```
============ 120 passed, 2 skipped, 5 xfailed in 167.61s (0:02:47) =============
Assertion 'e->n_sources == 0' failed at src/libsystemd/sd-event/sd-event.c:355, function event_free(). Aborting.
make: *** [Makefile:9016: pytest-cov] Aborted (core dumped)
```

There's zero google hits for that. @allisonkarlitskaya did you ever run into this by any chance?

martinpitt commented:
I have rebased this to get it back into working order. #19668 has been open for 7 months, and at some point I'd really like to unblock this and the follow-up work -- the string parsing hack is easy enough to revert/adjust if/once we land the FernyTransport rework.

martinpitt commented:
I'm a bit tired of this PR being blocked for so long, and due to $reasons I don't see #19668 land in the next 4 months either. That string parsing hack isn't so bad IMHO, it's effectively the same what ferny does (just admittedly less elegant), but both are covered by integration tests.  So I rebased this again, added a proper pointer to #19668, and ask for review again. Thanks!

martinpitt commented:
@allisonkarlitskaya gentle review ping?

allisonkarlitskaya commented:
> @allisonkarlitskaya gentle review ping?

Let's talk about this today in 1:1.

martinpitt commented on src/cockpit/beiboot.py:
```diff
@@ -181,9 +196,20 @@ def __init__(self, router: Router, destination: str, args: argparse.Namespace):
 
     async def do_connect_transport(self) -> None:
+        # do we have user/password (Basic auth) from the login page?
+        auth = await self.router.request_authorization("*")
```
@allisonkarlitskaya IIRC you were worried that this PR changes the Client. It's not supposed to, as that will just "fail" (i.e. give empty-ish reply `":"`) on the client, as the login page doesn't *have* user and password fields. This only applies to using cockpit-beiboot as cockpit-ssh replacement on the "normal" login page.

allisonkarlitskaya commented on src/cockpit/beiboot.py:
```diff
@@ -129,10 +129,25 @@ class AuthorizeResponder(ferny.AskpassHandler):
+
+        if self.have_basic_password and 'password:' in prompt.lower():
+            # the UI currently expects us to fail on a wrong password, so only send it once
+            if self.basic_password is not None:
```
This confused the heck out of me until I realized that you were setting `self.basic_password = None` below.

As far as I can tell, the code is correct as it's written, but the `have_basic_password` should definitely be renamed to something like `had_` or `got_` to remove the suggestion that it applies to the current state.

I might also just drop that and do something like

```python
if self.basic_password and 'password:' in prompt.lower():
    if self.password_attempted:
        raise ...

    self.password_attempted = True
    return self.basic_password
```

It has the disadvantage of not "freeing" the password, but you're not really scrubbing it anyway...

martinpitt replied:
Wrt. the renaming: it may be the current state, just that have_basic_password stays (to remember in which "mode" we are), and basic_password gets consumed, as it doesn't make sense to try more than once. `got_` is certainly fine, but if you think `password_attempted` is cleaner to read, that's fine for me.

Scrubbing in Python is indeed difficult, as `str` is immutable AFAIK. I.e. the only thing I can hope for is that the memory gets GCed and eventually overwritten. This may be possible with `StringIO` or similar, but I figure all of this is kind of hopeless in Python, and we rather rely on ptrace protection and such..

allisonkarlitskaya commented on src/cockpit/beiboot.py:
```diff
@@ -181,9 +196,20 @@ def __init__(self, router: Router, destination: str, args: argparse.Namespace):
+            user, basic_password = base64.b64decode(auth.split(' ', 1)[1]).decode().split(':', 1)
+            if user:  # this can be empty, i.e. auth is just ":"
+                logger.debug("got username %s and password from Basic auth", user)
+                ssh_opts = ['-l', user]
```
What happens if you give a username and connect to a destination like `admin@c`.  Who wins?  Should we have a test?

martinpitt replied:
The user name wins: `ssh -l admin root@c` logs in as "admin". I did that in commit 38fa110a020c for the shell, happy to do a similar test case for the login page.

martinpitt replied:
Correction: We already do have a test, right in this PR:
```py
        # colliding usernames; user names in "Connect to:" are *not* supported,
        # but pin down the behaviour
        try_login("admin", "foobar", server=f"other@{my_ip}")
       check_session()
```

allisonkarlitskaya replied:
Cool!  Didn't see that, thanks!

martinpitt commented on pkg/static/login.js:
```diff
@@ -747,15 +750,48 @@ import "./login.scss";
+
+                    // otherwise, the key changed, let user confirm
+                } else {
+                    // cockpit-ssh and ssh disagree no which key type to show to the user
```
s/no/on/

martinpitt replied:
fixed

martinpitt commented on test/verify/check-static-login:
```diff
@@ -1010,6 +1010,28 @@ Command = /usr/bin/env python3 -m cockpit.beiboot
+        # login treats this as new host
+        try_login("admin", "foobar", expect_hostkey=True)
+        check_session()
+
```
for completeness, let's test changed host key/fingerprint as well

martinpitt replied:
added

allisonkarlitskaya commented on pkg/static/login.js:
```diff
@@ -1,5 +1,8 @@
 import "./login.scss";
 
+import sha256 from "js-sha256";
+import { base64_decode, base64_encode } from "_internal/base64";
```
lol that didn't take long...

allisonkarlitskaya reviewed (commented):
A thought: in the browser we know what key type we already know for the host in question.  Could we somehow proactively send that to the -ws in a way that gets passed on to `cockpit.beiboot` to either let it (a) do some more intelligent comparison, or (b) hint ssh about which kind of host key we want to see?

Proper review coming tomorrow :)

martinpitt commented:
Wrt. the active key type request: This is a problem which we've had forever, in main: The login page only receives (and stores) a single key in its database, whichever key type ssh decides to use and report. Aside from the fact that we want to get rid of cockpit-ssh, `cockpitsshrelay.c` also doesn't seem to have a lot of knobs for requesting additional host key types from the server -- the structure of the dialog is pretty much given by libssh.

What could possibly work is to (1) get rid of cockpit-ssh and replace it with cockpit-beiboot/ssh (for which this PR is the first stepping stone, but there's still a lot of further work to do), (2) let cockpit-beiboot call ssh(1) in a way that it writes the host keys into a temporary `known_hosts` file (or fish it out from the existing one -- this is all very handwavy and probably hard work), and (3) set up some sync between that file and localStore.

TBH this is the *very last* thing I want to work on now...

allisonkarlitskaya commented on pkg/static/login.js:
```diff
@@ -747,15 +750,48 @@ import "./login.scss";
+            * "myhost ecdsa-sha2-nistp256 AAAA[..]ToM=\n"; ssh(1) does not show them, only fingerprints, so
+            * cockpit-beiboot only responds with fingerprints as well; they look like "myhost ED25519 SHA256:JVG[..]xs"
+            * check if we can match the given fingerprint to an existing entry */
+            const [db_key_type, db_key_value] = db_key.split(" ").slice(1);
```
I guess this could very easily return the wrong number of arguments...

martinpitt replied:
Yes it could, hence the code below needs to be careful to get along with an `undefined` db_key_value. That's why it checks for its truthiness first.

allisonkarlitskaya commented on pkg/static/login.js:
```diff
@@ -747,15 +750,48 @@ import "./login.scss";
+            * check if we can match the given fingerprint to an existing entry */
+            const [db_key_type, db_key_value] = db_key.split(" ").slice(1);
+            if (key_value.startsWith('SHA256:') && db_key_value && !db_key_value.includes(':')) {
+                // cockpit-ssh and ssh use different notation: ED25519 vs. ssh-ed25519
```
```c
⬢[lis@toolbox openssh-portable]$ git grep '\.name =\|\.shortname ='
...
ssh-dss.c:      /* .name = */           "ssh-dss",
ssh-dss.c:      /* .shortname = */      "DSA",
ssh-dss.c:      /* .name = */           "ssh-dss-cert-v01@openssh.com",
ssh-dss.c:      /* .shortname = */      "DSA-CERT",
ssh-ecdsa-sk.c: /* .name = */           "sk-ecdsa-sha2-nistp256@openssh.com",
ssh-ecdsa-sk.c: /* .shortname = */      "ECDSA-SK",
ssh-ecdsa-sk.c: /* .name = */           "sk-ecdsa-sha2-nistp256-cert-v01@openssh.com",
ssh-ecdsa-sk.c: /* .shortname = */      "ECDSA-SK-CERT",
ssh-ecdsa-sk.c: /* .name = */           "webauthn-sk-ecdsa-sha2-nistp256@openssh.com",
ssh-ecdsa-sk.c: /* .shortname = */      "ECDSA-SK",
ssh-ecdsa.c:    /* .name = */           "ecdsa-sha2-nistp256",
ssh-ecdsa.c:    /* .shortname = */      "ECDSA",
ssh-ecdsa.c:    /* .name = */           "ecdsa-sha2-nistp256-cert-v01@openssh.com",
ssh-ecdsa.c:    /* .shortname = */      "ECDSA-CERT",
ssh-ecdsa.c:    /* .name = */           "ecdsa-sha2-nistp384",
ssh-ecdsa.c:    /* .shortname = */      "ECDSA",
ssh-ecdsa.c:    /* .name = */           "ecdsa-sha2-nistp384-cert-v01@openssh.com",
ssh-ecdsa.c:    /* .shortname = */      "ECDSA-CERT",
ssh-ecdsa.c:    /* .name = */           "ecdsa-sha2-nistp521",
ssh-ecdsa.c:    /* .shortname = */      "ECDSA",
ssh-ecdsa.c:    /* .name = */           "ecdsa-sha2-nistp521-cert-v01@openssh.com",
ssh-ecdsa.c:    /* .shortname = */      "ECDSA-CERT",
ssh-ed25519-sk.c:       /* .name = */           "sk-ssh-ed25519@openssh.com",
ssh-ed25519-sk.c:       /* .shortname = */      "ED25519-SK",
ssh-ed25519-sk.c:       /* .name = */           "sk-ssh-ed25519-cert-v01@openssh.com",
ssh-ed25519-sk.c:       /* .shortname = */      "ED25519-SK-CERT",
ssh-ed25519.c:  /* .name = */           "ssh-ed25519",
ssh-ed25519.c:  /* .shortname = */      "ED25519",
ssh-ed25519.c:  /* .name = */           "ssh-ed25519-cert-v01@openssh.com",
ssh-ed25519.c:  /* .shortname = */      "ED25519-CERT",
ssh-rsa.c:      /* .name = */           "ssh-rsa",
ssh-rsa.c:      /* .shortname = */      "RSA",
ssh-rsa.c:      /* .name = */           "ssh-rsa-cert-v01@openssh.com",
ssh-rsa.c:      /* .shortname = */      "RSA-CERT",
ssh-rsa.c:      /* .name = */           "rsa-sha2-256",
ssh-rsa.c:      /* .shortname = */      "RSA",
ssh-rsa.c:      /* .name = */           "rsa-sha2-512",
ssh-rsa.c:      /* .shortname = */      "RSA",
ssh-rsa.c:      /* .name = */           "rsa-sha2-256-cert-v01@openssh.com",
ssh-rsa.c:      /* .shortname = */      "RSA-CERT",
ssh-rsa.c:      /* .name = */           "rsa-sha2-512-cert-v01@openssh.com",
ssh-rsa.c:      /* .shortname = */      "RSA-CERT",
ssh-xmss.c:     /* .name = */           "ssh-xmss@openssh.com",
ssh-xmss.c:     /* .shortname = */      "XMSS",
ssh-xmss.c:     /* .name = */           "ssh-xmss-cert-v01@openssh.com",
ssh-xmss.c:     /* .shortname = */      "XMSS-CERT",
```

martinpitt replied:
Meh.. But I just realized that we don't really actually have to do this comparison. If we assume collision resistance (SHA256 's only job), then we can either be much more lenient about this and do a lower-case substring match, or even just  drop it completely -- the hash of a different key type is going to be different. If not, the world has much bigger problems :cold_sweat:

martinpitt replied:
This comparison is gone now.

martinpitt replied:
There's no fingerprint conversion or all this type magic any more.

allisonkarlitskaya commented on pkg/static/login.js:
```diff
@@ -747,15 +750,48 @@ import "./login.scss";
+            if (key_value.startsWith('SHA256:') && db_key_value && !db_key_value.includes(':')) {
+                // cockpit-ssh and ssh use different notation: ED25519 vs. ssh-ed25519
+                if (key_type.replace('ssh-', '').toLowerCase() == db_key_type.replace('ssh-', '').toLowerCase()) {
+                    // cockpit-ssh and ssh agree no which key type to show to the user;
```
agree *on

martinpitt replied:
The comment is gone now.

allisonkarlitskaya commented on pkg/static/login.js:
```diff
@@ -747,15 +750,48 @@ import "./login.scss";
+                if (key_type.replace('ssh-', '').toLowerCase() == db_key_type.replace('ssh-', '').toLowerCase()) {
+                    // cockpit-ssh and ssh agree no which key type to show to the user;
+                    // compute SHA256 fingerprint of reference db key and compare it
+                    const fp_raw = 'SHA256:' + base64_encode(sha256.digest(base64_decode(db_key_value)));
```
I would never have thought that this was so straightforward...

allisonkarlitskaya commented on pkg/static/login.js:
```diff
@@ -747,15 +750,48 @@ import "./login.scss";
+                    // cockpit-ssh and ssh disagree on which key type to show to the user
+                    console.warn("host key verification: host sent us key type", key_type,
+                                 "but our localStore database has type", db_key_type, "; treating as new host");
+                    db_key = null;
```
I mentioned on IRC why I think this is a bad decision.

If someone wants to attack a user, and they know that this user uses Cockpit and recently went through this upgrade step then they can tune their attack by MITM-ing to a server which presents a key of a different type than the one that the user has in their DB in order that they get a more-innocent looking "new host" dialog instead of the "you are being attacked!!" one.  ie: in this scenario it's not a case of cockpit-ssh and ssh disagreeing, it's a case of the attacker having manipulated the scenario.

If we further assume that our attackers will always take advantage of this possibility to elicit a less-scary dialog, then it means that we'll actually only see the super-scary dialog in the case where we're *not* under attack (ie: key change for innocent reasons).

So ya: I don't like that we give the attacker the trivial ability to downgrade the type of warning.

martinpitt replied:
I think I misunderstood you. This situation can happen all the time, even with `ssh` itself. Over time and different server or libssh versions, the presented key type probably changed a lot. This PR doesn't fundamentally change that. But reading your full comment here I think that you merely want to show the "host key changed" dialog in this situation, which is fair (and I agree).

martinpitt replied:
Changed. This makes the patch much simpler actually, and it's more consistent. Sorry for the misunderstanding!

allisonkarlitskaya commented on src/cockpit/beiboot.py:
```diff
@@ -150,12 +166,27 @@ async def do_askpass(self, messages: str, prompt: str, hint: str) -> Optional[st
+            # we don't have acces to the full key, but the fingerprint is good enough
+            args['host-key'] = f'{host_match.group(1)} {fp_match.group(1)} {fp_match.group(2)}'
+            # very oddly named, login.js do_hostkey_verification() expects the fingerprint here
+            args['default'] = fp_match.group(2)
```
It seems a bit weird that we make this the default.  Is that something that the current dialog expects/requires?

martinpitt replied:
Yes, see the comment above it. With cockpit-ssh, `default` is the fingerprint while `host-key` is the full thing.

allisonkarlitskaya reviewed (commented):
I'm still quite worried about the migration.

And: once we do this, there's no going back.  If we want to imagine a future scenario where we put the key from the browser into a temporary known-hosts file and run `ssh` with that file and send the result (with added host keys) back to the browser, we can't do that anymore (since we only have fingerprints now, and hashing is one way).

I also know you want to get this done and out of the way so you can work on more interesting things...

Maybe I'm too worried about localstorage.  This is browsers after all.  People clear their cookies and lose their phones and stuff all the time, don't they?

martinpitt commented on pkg/static/login.js:
```diff
@@ -747,15 +750,48 @@ import "./login.scss";
+        if (db_key) {
+            // exact same key → good
+            if (key_db[key_key] == key) {
+                converse(data.id, data.default);
```
Yes, it is. It is really poorly named, but I don't want to change protocol here between three moving parts (login.js, cockpit-ssh, and cockpit-beiboot). Reworking this gets easier when at least cockpit-ssh is gone.

martinpitt commented:
> once we do this, there's no going back. If we want to imagine a future scenario where we put the key from the browser into a temporary known-hosts file and run ssh with that file and send the result (with added host keys) back to the browser, we can't do that anymore

There is no principal reason for stopping us here. We can do the inverse of what this PR does and check if we already know the fingerprint of the host, before calling ssh with a temporary known_hosts file. (Note that you can't write files from login.js, this needs a ton of work in cockpit-ws to start supporting that -- I don't see us do that anytime soon).

martinpitt commented on pkg/static/login.js:
```diff
@@ -747,15 +750,44 @@
+            * "myhost ecdsa-sha2-nistp256 AAAA[..]ToM=\n"; ssh(1) does not show them, only fingerprints, so
+            * cockpit-beiboot only responds with fingerprints as well; they look like "myhost ED25519 SHA256:JVG[..]xs"
+            * check if we can match the given fingerprint to an existing entry */
+            const [db_key_type, db_key_value] = db_key.split(" ").slice(1);
```
Fixed.

allisonkarlitskaya commented:
> Note that you can't write files from login.js, this needs a ton of work in cockpit-ws to start supporting that -- I don't see us do that anytime soon

`login.js` can pass arbitrary information to cockpit-beiboot via ~`Authorize` HTTP headers which get turned into `authorize` messages in Cockpit protocol.  cockpit-beiboot can just use `tempfile.NamedTemporaryFile()` or so while setting up to call `ssh`.  The `ferny.Session()` way of doing this already set up an entire temporary directory for the control socket and so on.  I don't think there needs to be anything going on with `cockpit-ws` here...

allisonkarlitskaya commented:
ps: I realize that this essentially represents a radical rethinking of the way we handle host keys in the browser localstorage world, and I know you didn't sign up for this, and really want to work on getting the -ws into the container...

...but if it worked, wouldn't it be nice?

martinpitt commented:
I'm still not a fan. You can't get the full keys into a known_hosts file without actually ack'ing the fingerprint and doing the connection to a machine, so this is would be a two-step process: first attempt tries to connect, fails, presents the FP to the user. Second connection attempt then sends the acked FP to beiboot/ssh, connects, and writes the known_hosts file, which then needs to be sent to the UI *after the connection succeeded*, and needs to be correlated to the first attempt, and stored. login.js then needs to special-case this for cockpit-beiboot as cockpit-ssh cannot do that. I want to kill cockpit-ssh first, then we have more leeway to change the protocol.

Doing another Authorize *after* success is indeed a big change in the protocol, and I don't have enough of it in my head to say how difficult it will be.

But in the end I consider it not worth spending that time and effort on. The only thing that this PR changes its storing full keys → their SHA256 fingerprint. This is cryptographically sane enough. I don't want to block this for yet another year and turn it into another #19668.. (Plus, this PR doesn't actually change *anything* in the product, it's just the prerequisite)

allisonkarlitskaya commented:
> I'm still not a fan. You can't get the full keys into a known_hosts file without actually ack'ing the fingerprint and doing the connection to a machine, so this is would be a two-step process: first attempt tries to connect, fails, presents the FP to the user.

This is a very good point and I hadn't considered it.

martinpitt commented on src/cockpit/protocol.py:
```diff
@@ -237,7 +237,8 @@ async def request_authorization(
         if self.authorizations is None:
             self.authorizations = {}
-        cookie = str(uuid.uuid4())
+        cookie = str(self.next_auth_id)
```
I agree -- this is just for telling multiple parallel auth requests apart, it's not  a secret.

martinpitt commented on src/cockpit/beiboot.py:
```diff
@@ -212,60 +245,68 @@ def __init__(self, router: Router, destination: str, args: argparse.Namespace):
+
+    async def connect_from_flatpak(self) -> None:
+        # We want to run a python interpreter somewhere...
+        cmd, env = python_interpreter('cockpit-bridge')
```
The signature of python_interpreter() and flatpak_spawn() is weird -- it returns an "env" which is always empty. This makes this code hard to read/understand. I.e. this here is nothing more than an initialization of `env = []`, which is easier to understand.

allisonkarlitskaya replied:
The idea is that we have a bunch of these functions and they return a pair of argv/envp.  The next thing in the chain expects an envp (and needs to be able to handle it) so it has to come from somewhere, and this is the most natural place...

martinpitt replied:
Yeah, I got it -- I figure we should make all functions *receive* an env as well then, and return the concatenation.

martinpitt replied:
Not important, I left it as-is for now.

martinpitt commented on src/cockpit/beiboot.py:
```diff
@@ -212,60 +245,68 @@ def __init__(self, router: Router, destination: str, args: argparse.Namespace):
+        # Remote host?  Wrap command with SSH
+        if self.destination != 'localhost':
+            # we run ssh and thus the helper on the host, always use the xdg-cache helper
+            cmd, env = via_ssh(cmd, self.destination, ensure_ferny_askpass())
```
... and moreover, it would stop making this look like it'd entirely ignore the env from the python_interpreter() call above.

martinpitt commented on src/cockpit/beiboot.py:
```diff
@@ -212,60 +245,68 @@ def __init__(self, router: Router, destination: str, args: argparse.Namespace):
+                command='authorize', challenge='x-login-data', cookie='-', login_data={
+                    'user-known-hosts': user_known_hosts_file.read_text()
+                }
+            )
```
This doesn't fit here in "split flatpak and bastion code paths", it should be part of the next commit ("support for temporary UserKnownHostsFile")

allisonkarlitskaya replied:
rebase fail :)

martinpitt replied:
Moved.

martinpitt reviewed (commented):
Thanks @allisonkarlitskaya ! I had a first look at your extra commits. I'll rebase/massage them a bit, and then make use of x-login-data on the login page.

martinpitt commented:
@allisonkarlitskaya I've looked at this again. The timing is a bit weird here -- login.js `do_hostkey_verification()` gets called immediately after the initial ssh connection (before doing user auth) -- that can continue to write the fingerprint. But `login-data` is only set/sent after *successful* authentication. Which means several things:

  * we need to first write a placeholder into localStorage and then replace it with the real value after a successful login
  * if you specify the wrong user credentials, we'll *never* record the actual SSH key, as we never receive `login-data`. So the next attempt after typing a failed password it would ask you about the fingerprint again, which is ugly.
 * To get around that, we could first write the fingerprint into localStore (as placeholder), overwrite it with full keys from login-data later, and adjust do_hostkey_verification() to accept either fingerprint or full key.
 * However, it means that in `do_hostkey_verification()` we can *never* rely on full ssh keys, which kind of thwarts that whole idea.

I wonder if we can send an authorize challenge response from beiboot.py earlier, immediately after SSH writes the known host file. Can the login page pick this up somehow? This is all exploding into a big rewrite excercise again..

martinpitt commented:
I did some rebasing and cleanups, and added some integration tests that check that localStorage stores the full host key. That fails (as it currently stores the fingerprint). But I don't want this to get lost.

allisonkarlitskaya commented:
My idea was roughly that we always fall into one of three categories:
 - didn't connect to the host (or didn't see any host key for some other reason): don't do anything
 - successfully connected, got host key, but had wrong credentials: we still need to report the host key via the "close" message.  I agree that we're gonna need to up our game here vs. what the current patch does.  Perhaps a new field on the close message with the entire file content.
 - successfully connected and logged in: login-data metadata

martinpitt commented:
@allisonkarlitskaya I got the login.js logic under control. The main (documented and tested) quirk is now not remembering the new full host keys on a failed login attempt (see #20954). I made this a blocker of the PR which will make cockpit-beiboot the default (#19441), but I don't think it needs to block *this* PR -- I think beiboot's host key behaviour here is already a *lot* better than the status quo.

I went through the discussion threads here and closed the outdated/resolved ones, to better see what's left. There's nothing outstanding right now, but I'm sure you can come up with some more :grin:

cockpituous commented on pkg/static/login.js:
```diff
@@ -1,5 +1,10 @@
 import "./login.scss";
 
+function debug(...args) {
+    if (window.debugging == 'all' || window.debugging?.includes('login'))
+        console.debug('login:', ...args);
```
These 2 added lines are not executed by any test.

cockpituous replied:
These 2 added lines are not executed by any test.

cockpituous replied:
These 2 added lines are not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -552,8 +557,7 @@ import "./login.scss";
     function host_failure(msg) {
-        const host = id("server-field").value;
-        if (!host) {
+        if (!login_machine) {
```
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -590,17 +594,28 @@ import "./login.scss";
         if (user === "" && !environment.is_cockpit_client) {
             login_failure(_("User name cannot be empty"));
-        } else if (need_host() && id("server-field").value === "") {
+        } else if (need_host() && login_machine === "") {
```
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -770,25 +805,31 @@ import "./login.scss";
 
-        if (key_db[key_key] == key) {
+        // code path for old C cockpit-ssh, which doesn't set a known_hosts file in advance (like beiboot)
+        if (key_db[key_host] == key) {
+            debug("do_hostkey_verification: received key matches known_hosts database, auto-accepting fingerprint", data.default);
```
These 2 added lines are not executed by any test.

cockpituous replied:
These 2 added lines are not executed by any test.

cockpituous replied:
These 2 added lines are not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -1024,6 +1085,20 @@ import "./login.scss";
+                    console.debug("setup_localstorage(): updating known_hosts database for deferred host key for", login_data_host, ":", hostkey);
+                    set_known_hosts_db(key_db);
+                } else {
+                    console.error("login.js internal error: setup_localstorage() received a pending login-data host, but login-data does not contain known-hosts");
```
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

martinpitt commented:
Most failures were very shallow, fixed locally. The [f-41](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19401-8a5f5588-20240830-125535-fedora-41-other/log.html#155) one is the most interesting:

> denied  { read } for  pid=1904 comm="ssh" name="20-systemd-ssh-proxy.conf" dev="vda4" ino=13369 scontext=system_u:system_r:cockpit_ws_t:s0 tcontext=system_u:object_r:systemd_conf_t:s0 tclass=file permissive=0

It seems fair/fine to amend our policy accordingly.

cockpituous commented on pkg/static/login.js:
```diff
@@ -1024,6 +1086,20 @@ import "./login.scss";
+                    console.debug("setup_localstorage(): updating known_hosts database for deferred host key for", login_data_host, ":", hostkey);
+                    set_known_hosts_db(key_db);
+                } else {
+                    console.error("login.js internal error: setup_localstorage() received a pending login-data host, but login-data does not contain known-hosts");
```
This added line is not executed by any test.

martinpitt commented:
@allisonkarlitskaya FYI, I had to rework the "support for temporary UserKnownHostsFile" commit quite a bit. Bigger scope of the temp dir, being careful about *not* supplying the `UserKnownHostsfile=` option if it's not part of the `Authorize:` header data (as we *do* want to use the user's default file for Client) and so on.

cockpituous commented on pkg/static/login.js:
```diff
@@ -1,5 +1,10 @@
 import "./login.scss";
 
+function debug(...args) {
+    if (window.debugging === 'all' || window.debugging?.includes('login'))
+        console.debug('login:', ...args);
```
These 2 added lines are not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -552,8 +562,7 @@ import "./login.scss";
     function host_failure(msg) {
-        const host = id("server-field").value;
-        if (!host) {
+        if (!login_machine) {
```
This added line is not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -590,17 +599,28 @@ import "./login.scss";
         if (user === "" && !environment.is_cockpit_client) {
             login_failure(_("User name cannot be empty"));
-        } else if (need_host() && id("server-field").value === "") {
+        } else if (need_host() && login_machine === "") {
```
This added line is not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -759,36 +799,48 @@ import "./login.scss";
 
-        if (key_db[key_key] == key) {
+        // code path for old C cockpit-ssh, which doesn't set a known_hosts file in advance (like beiboot)
+        if (db_keys == key) {
+            debug("do_hostkey_verification: received key matches known_hosts database, auto-accepting fingerprint", data.default);
```
These 2 added lines are not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -931,6 +993,22 @@ import "./login.scss";
+                            } else {
+                                debug("send_login_request():", ssh_host_key_change_host, "changed key, but did not get an updated key from response");
+                            }
+                        } catch (ex) {
+                            console.error("Failed to parse response text as JSON:", xhr.responseText, ":", JSON.stringify(ex));
```
These 2 added lines are not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -1024,6 +1112,18 @@ import "./login.scss";
+                    console.debug("setup_localstorage(): updating known_hosts database for deferred host key for", login_data_host, ":", hostkey);
+                    set_hostkeys(login_data_host, hostkey);
+                } else {
+                    console.error("login.js internal error: setup_localstorage() received a pending login-data host, but login-data does not contain known-hosts");
```
This added line is not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -766,29 +799,36 @@ function debug(...args) {
 
-        if (key_db[key_host] == key) {
+        // code path for old C cockpit-ssh, which doesn't set a known_hosts file in advance (like beiboot)
+        if (db_keys == key) {
```
This added line is not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -945,6 +993,22 @@ function debug(...args) {
+                            } else {
+                                debug("send_login_request():", ssh_host_key_change_host, "changed key, but did not get an updated key from response");
+                            }
+                        } catch (ex) {
+                            console.error("Failed to parse response text as JSON:", xhr.responseText, ":", JSON.stringify(ex));
```
These 2 added lines are not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -1038,6 +1112,18 @@ function debug(...args) {
+                    console.debug("setup_localstorage(): updating known_hosts database for deferred host key for", login_data_host, ":", hostkey);
+                    set_hostkeys(login_data_host, hostkey);
+                } else {
+                    console.error("login.js internal error: setup_localstorage() received a pending login-data host, but login-data does not contain known-hosts");
```
This added line is not executed by any test.

martinpitt commented:
[This](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19401-9e092269-20240902-132607-fedora-40-other/log.html#143) is a new failure due to the "Don't retry bad password three times" commit.  Apparently `NumberOfPasswordPrompts=1` also affects the passphrase prompts for the encrypted key. This is an artifact as we test client mode technically with "bastion host" mode, as there's no flatpak involved. It's probably best to just adjust the test accordingly, and I'll manually ensure that the real Client flatpak continues to ask three times.

martinpitt commented on src/cockpit/beiboot.py:
```diff
@@ -324,54 +335,69 @@ def do_init(self, message):
-        capabilities = message.setdefault('capabilities', {})
-        if not isinstance(capabilities, dict):
-            bridge.write_control(command='init', problem='protocol-error', message='capabilities must be a dict')
+    with tempfile.TemporaryDirectory() as tmpdir:
+        known_hosts_file = Path(tmpdir) / 'user-known-hosts'
```
@allisonkarlitskaya This large indented `with` block is ugly. I know how much you love contexts :grin: I'm also happy to move the temp dir creation into `__init__`, then we can drop that large indentation again.

allisonkarlitskaya replied:
I love contexts but I don't love large blocks of excessive indentation (or noisy patches).  Feel free to hang this on the `SshPeer`, instead...

martinpitt replied:
Moved to SshPeer.

cockpituous commented on pkg/static/login.js:
```diff
@@ -766,29 +798,36 @@ function debug(...args) {
 
-        if (key_db[key_host] == key) {
+        // code path for old C cockpit-ssh, which doesn't set a known_hosts file in advance (like beiboot)
+        if (db_keys == key) {
```
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -945,6 +992,22 @@ function debug(...args) {
+                            } else {
+                                debug("send_login_request():", ssh_host_key_change_host, "changed key, but did not get an updated key from response");
+                            }
+                        } catch (ex) {
+                            console.error("Failed to parse response text as JSON:", xhr.responseText, ":", JSON.stringify(ex));
```
These 2 added lines are not executed by any test.

cockpituous replied:
These 2 added lines are not executed by any test.

cockpituous replied:
These 2 added lines are not executed by any test.

cockpituous replied:
These 2 added lines are not executed by any test.

cockpituous replied:
These 2 added lines are not executed by any test.

cockpituous replied:
These 2 added lines are not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -1038,6 +1111,18 @@ function debug(...args) {
+                    console.debug("setup_localstorage(): updating known_hosts database for deferred host key for", login_data_host, ":", hostkey);
+                    set_hostkeys(login_data_host, hostkey);
+                } else {
+                    console.error("login.js internal error: setup_localstorage() received a pending login-data host, but login-data does not contain known-hosts");
```
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

martinpitt commented:
Finally! That was a lot of corner cases, at least an icosahedron :wink: 

I also updated #19441 which will put it all together, but will require some more design discussion (see its TODO list). But it's useful to see where this will lead. I'll debug some failures tomorrow, but #19441 will be a post-holiday thing for me.

allisonkarlitskaya commented on test/verify/check-client:
```diff
@@ -241,6 +241,17 @@ Command = /usr/bin/env python3 -m cockpit.beiboot
         b.click("#login-button")
         b.wait_val("#conversation-input", "")
+
+        # quirk: this test really runs the "bastion host" code path in beiboot, as we are not actually
```
I'd prefer if we hack the filesystem from the test case to make this run the flatpak code path instead...

allisonkarlitskaya commented on src/cockpit/beiboot.py:
```diff
@@ -130,10 +130,25 @@ class AuthorizeResponder(ferny.AskpassHandler):
+                self.basic_password = None
+                return reply
+            else:
+                logger.debug("AuthorizeResponder: prompt %r, used up our password, failing", prompt)
```
Since we only expect one password prompt, this should be more like an `assert` now, no?

martinpitt replied:
Right, I think this was from before the `NumPasswordPrompts=1` fix. I strengthened this now.

allisonkarlitskaya commented on src/cockpit/beiboot.py:
```diff
@@ -130,10 +130,25 @@ class AuthorizeResponder(ferny.AskpassHandler):
+                return reply
+            else:
+                logger.debug("AuthorizeResponder: prompt %r, used up our password, failing", prompt)
+                return messages.strip().splitlines()[-1]
```
What is this trying to do, exactly?

martinpitt replied:
It returns the last line of ssh (which is usually something like "wrong password blabla") to the UI via becoming the `message` bit of the generated CockpitError in the `except ferny.InteractionError` interpretation in `run()`. I previously raised an `authentication-failed` CockpitError directly, but you asked me to convert it to stderr.

martinpitt replied:
See previous comment, turns out we don't need this any more.

allisonkarlitskaya commented on src/cockpit/beiboot.py:
```diff
@@ -234,17 +236,22 @@ async def connect_from_flatpak(self) -> None:
-            user, basic_password = base64.b64decode(response.split(' ', 1)[1]).decode().split(':', 1)
+            decoded = base64.b64decode(response[6:]).decode()
+            user_password, sep, data = decoded.partition('\0')
+            if sep:
```
Note to self: `'\0'` is truthy.

allisonkarlitskaya commented on src/cockpit/beiboot.py:
```diff
@@ -256,7 +263,11 @@ async def connect_from_bastion_host(self) -> None:
             logger.error("Could not find cockpit-askpass helper at %r", askpass)
 
-        cmd, env = via_ssh(cmd, self.destination, ssh_askpass, *username_opts)
+        if known_hosts is not None:
```
This feels slightly wrong.  I think we always want to set this.  Even if the user didn't specify it to us in the first place, we will want to have the tempfile there to report results back to the user.

I'm guessing the `is None` path is never actually hit since whenever you run this code the login page is unconditionally sending us `'\0'` in our auth request?  This could use some clarification in any case...

btw: it would be nice to move this stuff to ferny eventually.

martinpitt replied:
I started with always setting it, but that's not compatible with the old login page which did not set it (that caused several test failures, and was one of the corner cases which took me a long time to get right). But moreover, we don't set this in Client mode, and we really really *don't* want to set a custom temp knownhosts file then -- it should use the user's real `~/.ssh/known_hosts`.

martinpitt replied:
I added a big comment explaining it.

allisonkarlitskaya replied:
This is inside of `connect_from_bastion_host()` — it's not run in the Cockpit Client case.

Also, I don't think we have to worry about compatibility with the old login page.  `cockpit-ws`, `pkg/static/` and `cockpit-beiboot` will always be shipped together.

allisonkarlitskaya replied:
It's an interesting point about the tmpdir scope, though: we shouldn't create the tempdir when we're in the flatpak case.  So don't move that to `__init__()`....

martinpitt replied:
> it's not run in the Cockpit Client case.

Good point -- however, it still feels wrong to me to hide the user's ~/.ssh/known_hosts if the caller didn't explicitly ask for it. With the recent test adjustments to TestClient etc. it'll probably work, but it does feel unclean to me. If you prefer, I can change it though.

> we shouldn't create the tempdir when we're in the flatpak case.

Doesn't hurt much, does it? Current push has it in __init__, but I can move it back of course (no strong opinion)

martinpitt replied:
TestClient and testBeibootWithBridge both pass with this:

```diff
--- src/cockpit/beiboot.py
+++ src/cockpit/beiboot.py
@@ -279,12 +279,7 @@ class SshPeer(Peer):
 
         if response.startswith('Basic '):
             decoded = base64.b64decode(response[6:]).decode()
-            user_password, sep, data = decoded.partition('\0')
-            # be careful: not specifying known-hosts data at all is different from an empty set of keys
-            # the former means "use user's default ~/.ssh/config" (for Client), the latter uses a temporary file
-            # (for bastion host mode of the login page)
-            if sep:
-                known_hosts = data
+            user_password, _, known_hosts = decoded.partition('\0')
             user, _, basic_password = user_password.partition(':')
             if user:  # this can be empty, i.e. auth is just ":"
                 logger.debug("got username %s and password from Basic auth", user)
```

So it would *work*, but I still don't prefer it -- I'd rather be a bit more careful here.

allisonkarlitskaya replied:
Here's my take: we have two (and a half) scenarios that this code will run in:
 - flatpak for Cockpit Client (will not run this function)
 - cockpit-ssh for bastion host: in this case we're actually running ssh as the cockpit-wsinstance user on behalf of someone logging in via the web and we never ever want it to have access to `~/.ssh`, which hopefully doesn't even exist.  We absolutely always want to create an isolated environment for this user.  (mental note: maybe we should drop privs in this case...)
   - (the half case) same as above, but inside of a container

martinpitt replied:
OK -- I applied the above patch.

allisonkarlitskaya replied:
But what's the scenario where we accept that we didn't receive a `Basic` auth, then?  You still allow that `known_hosts` may be None, which can now only happen if we skip that entire block...

Recall: this scenario is only for bastion host.  I think we should *always* have a Basic auth in that case, no?

I could have probably done some more simplification when I split these two paths... but this commit ended up coming 'after' in the series...

martinpitt replied:
As I said, it will work (tests are green). I don't know if there's third-party scenarios which speak that protocol (foreman? etc.). But we'll know :grin: And these should also not use ~/.ssh/

allisonkarlitskaya reviewed (changes requested):
Mostly just read the Python bits so far, without reading the login page.  This is looking really good — lots of progress in the last 24 hours.

A few comments.

martinpitt commented:
I figured out the "invalid conversation token" bug which broke [this](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19441-c34aa521-20240902-171157-fedora-40-firefox-other/log.html) and [this](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19441-c34aa521-20240902-171206-fedora-40-firefox-expensive/log.html) (see #19441). Commit added here.

allisonkarlitskaya commented on src/cockpit/beiboot.py:
```diff
@@ -151,14 +152,19 @@ async def do_askpass(self, messages: str, prompt: str, hint: str) -> Optional[st
             return None
 
-        challenge = 'X-Conversation - ' + base64.b64encode(prompt.encode()).decode()
+        challenge_id = f'{os.getpid()}-{time.time()}'
```
Can we make this a simple counter to make it easier to chat with this thing via `cockpit.print`?

martinpitt replied:
No, I tried that first. as beiboot gets invoked multiple times, and then always restarts with 0.

allisonkarlitskaya replied:
Do we gain something by verifying the response, though?  We already matched this via the cookie...

martinpitt replied:
Not much, just additional confidence. It's mostly a safeguard in case the routing fails, to not take a non-matching response.

martinpitt commented:
@allisonkarlitskaya do you still want to review this, or are you firmly outside of cockpit land now? :grin: (that's fine -- we were by and large d'accord on this, Marius could review the rest)

martinpitt commented:
@mvollmer Can you please review this? This PR almost celebrates its first birthday, and I sorted out all the disagreements with @allisonkarlitskaya. This is a prerequisite for fully eliminating cockpit-ssh in PR #19441. That mostly works, but still needs some more things done.

martinpitt commented:
Just rebasing to current main to get an updated set of tests, as this became stale. No actual changes.

jelly commented on test/verify/check-static-login:
```diff
@@ -1012,6 +1024,35 @@ Command = /usr/bin/env python3 -m cockpit.beiboot
+        b.eval_js(f"""window.localStorage.setItem('known_hosts', '{{"{my_ip}": "{bad_key}"}}')""")
+        try_login("admin", "foobar")
+        b.wait_text("#hostkey-title", f"{my_ip} key changed")
+        b.wait_in_text("#hostkey-fingerprint", "SHA256:")
```
Do we want to bother testing ED25519 host keys?

martinpitt replied:
We actually do -- our VMs use that by default. But we don't really care about the specific key type anywhere -- login page and localStorage treat that as an opaque value.

jelly replied:
Was more about the custom regex parsing. I think we actually don't test IPv6 in this scenario.

martinpitt commented on src/cockpit/beiboot.py:
```diff
@@ -210,6 +211,8 @@ class SshPeer(Peer):
         self.destination = destination
         self.always = args.always
+        self.tmpdir = tempfile.TemporaryDirectory()
+        self.known_hosts_file = Path(self.tmpdir.name) / 'user-known-hosts'
```
In the bastion host case it is the initial thing launched by cockpit-ws (instead of cockpit-ssh). Keeping the known hosts file around throughout its lifetime seems fine to me?

For the host switcher it doesn't run, as remote.py uses its own `SshPeer` implemenation -- it doesn't even import beiboot.py (this is meant to be a top-level executable).

jelly replied:
If the bastion host has multiple users, do they use the same `beiboot` instance? And if not that would require some locking?

martinpitt replied:
No, every user *session* (even for the same user) gets its own cockpit-beiboot instance. It's the top-level session program (similar to `cockpit-session`).

jelly replied:
Ok, then this is resolved.

martinpitt commented on src/cockpit/beiboot.py:
```diff
@@ -167,14 +168,30 @@ async def do_askpass(self, messages: str, prompt: str, hint: str) -> Optional[st
 
+        # FIXME: is this a host key prompt? This should be handled more elegantly,
+        # see https://github.com/cockpit-project/cockpit/pull/19668
+        fp_match = re.search(r'\n(\w+) key fingerprint is ([^.]+)\.', prompt)
```
It's called once for every click on "Log in" button on the login page. The reason is in the commit message:

> We can't use ferny's builtin `do_hostkey()` responder for this, as that
> requires `ferny.Session(handle_host_key=True)`, and that API is not
> flexible enough to handle our ssh command modifications and the extra
> beiboot_helper handler. This needs some bigger redesign, see #19668.

It's got little to do with RHEL 8 (although that makes it even harder, it doesn't support the `KnownHostsCommand` -- i.e. ferny doesn't do `do_hostkey()` on RHEL 8 either).

jelly replied:
Yes, that was my point, RHEL-8 does not do do_hostkey and thus had that conditional.

jelly commented on src/cockpit/beiboot.py:
```diff
@@ -131,10 +131,25 @@ class AuthorizeResponder(ferny.AskpassHandler):
+    def __init__(self, router: Router, basic_password: Optional[str]):
         self.router = router
+        self.basic_password = basic_password
+        self.have_basic_password = bool(basic_password)
```
This would I presume break auth when a user has an empty password, I am not sure if that is supposed to be supported. That is my point, because:

`bool("") => False`

martinpitt replied:
Ah, I see -- I suppose I could add a test case for an empty password (not sure if that works with ssh), and then do the above condition as "user **and** password are empty → user_password = None". I don't know if it's supported as we never tested that anywhere.

martinpitt replied:
It works with ssh with

    echo "PermitEmptyPasswords yes" > /etc/ssh/sshd_config.d/01-empty-password.conf

And then I can also log into Cockpit with beiboot. 

This works because there isn't any actual prompt from ssh for a password, it logs you straight in.

So `""` never actually appears in this code path, and thus `is not None` just works as well. Changed accordingly.

martinpitt commented:
@jelly Addressed your comments: [interdiff](https://github.com/cockpit-project/cockpit/compare/afce95b08773630e4d9f16e86a22e03f3bb0bc21..a33575dc9ff4a938f8ac83d9a2cfa03b302cd925)

cockpituous commented on pkg/static/login.js:
```diff
@@ -766,29 +802,36 @@ function debug(...args) {
 
-        if (key_db[key_host] == key) {
+        // code path for old C cockpit-ssh, which doesn't set a known_hosts file in advance (like beiboot)
+        if (db_keys == key) {
```
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -945,6 +996,22 @@ function debug(...args) {
+                            } else {
+                                debug("send_login_request():", ssh_host_key_change_host, "changed key, but did not get an updated key from response");
+                            }
+                        } catch (ex) {
+                            console.error("Failed to parse response text as JSON:", xhr.responseText, ":", JSON.stringify(ex));
```
These 2 added lines are not executed by any test.

cockpituous replied:
These 2 added lines are not executed by any test.

cockpituous replied:
These 2 added lines are not executed by any test.

cockpituous commented on pkg/static/login.js:
```diff
@@ -1038,6 +1115,18 @@ function debug(...args) {
+                    console.debug("setup_localstorage(): updating known_hosts database for deferred host key for", login_data_host, ":", hostkey);
+                    set_hostkeys(login_data_host, hostkey);
+                } else {
+                    console.error("login.js internal error: setup_localstorage() received a pending login-data host, but login-data does not contain known-hosts");
```
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

cockpituous replied:
This added line is not executed by any test.

martinpitt commented:
Oh-uh, setting an empty password with `chpasswd` [does not work on Debian](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19401-6699c33d-20240920-164133-debian-testing-other/log.html#155) or [arch](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19401-6699c33d-20240920-164131-arch-other/log.html#155).

There's a related problem on [rhel-9-5](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19401-6699c33d-20240920-164141-rhel-9-5-other/log.html#155) where `chpasswd` works, but ssh refuses it.

Fixed both by using `passwd -d` (TIL!)

martinpitt commented:
Interesting [arch bug](https://cockpit-logs.us-east-1.linodeobjects.com/pull-19401-242e2d37-20240920-175007-arch-other/log.html) -- after `passwd -d admin`, `echo 'admin:foobar' | chpasswd` succeeds, and /etc/shadow does have a password again -- but it doesn't work. Changing it with `passwd admin` doesn't work either. Huh?
